// sqlit/dist/misc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const PLURAL_FORMS = {
    child: 'children'
};
function pluralise(name) {
    for (const key in PLURAL_FORMS) {
        if (name.endsWith(key)) {
            return name.substr(0, name.length - key.length) + PLURAL_FORMS[key];
        }
        if (name.endsWith(_U(key))) {
            return name.substr(0, name.length - key.length) + _U(PLURAL_FORMS[key]);
        }
    }
    let result;
    if ((result = name.replace(/([^aeiou])y$/i, '$1ies')) != name) {
        return result;
    }
    if ((result = name.replace(/ty$/, 'ties')) != name) {
        return result;
    }
    if ((result = name.replace(/s$/, 'ses')) != name) {
        return result;
    }
    return name + 's';
}
exports.pluralise = pluralise;
function _U(s) {
    return s.charAt(0).toUpperCase() + s.slice(1);
}
function setPluralForms(data) {
    for (const key in data) {
        PLURAL_FORMS[key] = data[key];
    }
}
exports.setPluralForms = setPluralForms;
function setPluralForm(singular, plural) {
    PLURAL_FORMS[singular] = plural;
}
exports.setPluralForm = setPluralForm;
function toCamelCase(s) {
    return s.replace(/_\w/g, m => m[1].toUpperCase());
}
exports.toCamelCase = toCamelCase;
function toPascalCase(s) {
    s = toCamelCase(s);
    return s[0].toUpperCase() + s.substr(1);
}
exports.toPascalCase = toPascalCase;
function toArray(args) {
    return Array.isArray(args) ? args : [args];
}
exports.toArray = toArray;
const reflect = p => p.then(value => ({ value }), error => ({ error }));
function promiseAll(promises) {
    return Promise.all(promises.map(reflect)).then(results => {
        const error = results.find(result => 'error' in result);
        if (error) {
            throw error.error;
        }
        else {
            return results.map(result => result.value);
        }
    });
}
exports.promiseAll = promiseAll;
