// sqlit/dist/engine/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const information_schema_1 = require("./information_schema");
exports.getInformationSchema = information_schema_1.getInformationSchema;
class QueryCounter {
    constructor() {
        this.total = 0;
    }
}
exports.QueryCounter = QueryCounter;
class Connection {
    commit() {
        return this.query('commit');
    }
    rollback() {
        return this.query('rollback');
    }
}
exports.Connection = Connection;
class ConnectionPool {
}
exports.ConnectionPool = ConnectionPool;
function createConnectionPool(dialect, connection) {
    if (dialect === 'mysql') {
        return require('./mysql').default.createConnectionPool(connection);
    }
    throw Error(`Unsupported engine type: ${dialect}`);
}
exports.createConnectionPool = createConnectionPool;
function createConnection(dialect, connection) {
    if (dialect === 'mysql') {
        return require('./mysql').default.createConnection(connection);
    }
    throw Error(`Unsupported engine type: ${dialect}`);
}
exports.createConnection = createConnection;
