// sqlit/dist/engine/sqlite3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _1 = require(".");
const sqlite3 = require("sqlite3");
class _ConnectionPool extends _1.ConnectionPool {
    constructor(options) {
        super();
        this.options = options;
    }
    getConnection() {
        return new Promise((resolve, reject) => {
            const connection = new sqlite3.Database(this.options.database, undefined, error => {
                if (error)
                    reject(Error(error));
                resolve(new _Connection(connection, true));
            });
        });
    }
    close() {
        return Promise.resolve();
    }
    escape(value) {
        return escapeString(value);
    }
    escapeId(name) {
        return escapeId(name);
    }
}
class _Connection extends _1.Connection {
    constructor(options, connected) {
        super();
        this.dialect = 'sqlite3';
        this.queryCounter = new _1.QueryCounter();
        if (connected) {
            this.connection = options;
        }
        else {
            this.connection = new sqlite3.Database(options.database);
        }
    }
    release() {
        this.disconnect();
    }
    query(sql) {
        this.queryCounter.total++;
        return new Promise((resolve, reject) => this.connection.query(sql, (error, results, fields) => {
            if (error) {
                return reject(error);
            }
            if (Array.isArray(results)) {
                resolve(results);
            }
            else if (results.insertId) {
                resolve(results.insertId);
            }
            else {
                resolve(results.changedRows);
            }
        }));
    }
    transaction(callback) {
        return new Promise((resolve, reject) => {
            return this.connection.beginTransaction(error => {
                if (error)
                    return reject(error);
                let promise;
                try {
                    promise = callback(this);
                }
                catch (error) {
                    return this.connection.rollback(() => {
                        reject(error);
                    });
                }
                if (promise instanceof Promise) {
                    return promise
                        .then(result => this.connection.commit(error => {
                        if (error) {
                            return this.connection.rollback(() => {
                                reject(error);
                            });
                        }
                        else {
                            resolve(result);
                        }
                    }))
                        .catch(reason => this.connection.rollback(() => {
                        reject(reason);
                    }));
                }
                else {
                    resolve();
                }
            });
        });
    }
    commit() {
        return new Promise((resolve, reject) => {
            this.connection.commit(error => {
                if (error)
                    reject(error);
                else
                    resolve();
            });
        });
    }
    rollback() {
        return new Promise((resolve, reject) => {
            this.connection.rollback(error => {
                if (error)
                    reject(error);
                else
                    resolve();
            });
        });
    }
    disconnect() {
        return new Promise((resolve, reject) => {
            this.connection.close(err => {
                if (err)
                    reject(err);
                else
                    resolve();
            });
        });
    }
    escape(value) {
        return escapeString(value);
    }
    escapeId(name) {
        return escapeId(name);
    }
}
function escapeString(value) {
    return `'${value.replace(/'/g, "''")}'`;
}
function escapeId(name) {
    return `'${name}'`;
}
exports.default = {
    createConnectionPool: (options) => {
        return new _ConnectionPool(options);
    },
    createConnection: (options) => {
        return new _Connection(options);
    }
};
